File: HW01.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

// Base class for Inventory system
class InventorySystem {
protected:
    int productID;
    string name;
    float price;
    int amount_in_inventory;

public:
    InventorySystem(int productID, string name, float price, int amount_in_inventory)
        : productID(productID), name(name), price(price), amount_in_inventory(amount_in_inventory) {}

    virtual ~InventorySystem() {
        // Destructor implementation (if needed)
    }

    virtual void display_info() {
        cout << "Product ID: " << productID << ", Name: " << name << ", Price: $" << price
             << ", Available: " << amount_in_inventory << endl;
    }

    void buy_item(int quantity) {
        if (amount_in_inventory >= quantity) {
            amount_in_inventory -= quantity;
            cout << "Purchased " << quantity << " " << name << "(s)." << endl;
        } else {
            cout << "Not enough stock for " << name << ". Only " << amount_in_inventory << " available." << endl;
        }
    }

    void restock_item(int quantity) {
        amount_in_inventory += quantity;
        cout << "Restocked " << quantity << " " << name << "(s). Now available: " << amount_in_inventory << "." << endl;
    }

    int getProductID() const {
        return productID;
    }
};

// Derived class for Indoor Sports category
class IndoorSports : public InventorySystem {
public:
    IndoorSports(int productID, string name, float price, int amount_in_inventory)
        : InventorySystem(productID, name, price, amount_in_inventory) {}
};

// Derived class for Outdoor Sports category
class OutdoorSports : public InventorySystem {
public:
    OutdoorSports(int productID, string name, float price, int amount_in_inventory)
        : InventorySystem(productID, name, price, amount_in_inventory) {}
};

// Volleyball subclass
class Volleyball : public IndoorSports {
private:
    string type_of_item;
public:
    Volleyball(int productID, string name, float price, int amount_in_inventory, string type_of_item)
        : IndoorSports(productID, name, price, amount_in_inventory), type_of_item(type_of_item) {}

    void display_info() override {
        InventorySystem::display_info();
        cout << "Type: " << type_of_item << endl;
    }
};

// Basketball subclass
class Basketball : public IndoorSports {
private:
    string type_of_item;
public:
    Basketball(int productID, string name, float price, int amount_in_inventory, string type_of_item)
        : IndoorSports(productID, name, price, amount_in_inventory), type_of_item(type_of_item) {}

    void display_info() override {
        InventorySystem::display_info();
        cout << "Type: " << type_of_item << endl;
    }
};

// TableTennis subclass
class TableTennis : public IndoorSports {
private:
    string type_of_item;
public:
    TableTennis(int productID, string name, float price, int amount_in_inventory, string type_of_item)
        : IndoorSports(productID, name, price, amount_in_inventory), type_of_item(type_of_item) {}

    void display_info() override {
        InventorySystem::display_info();
        cout << "Type: " << type_of_item << endl;
    }
};

// Camping subclass
class Camping : public OutdoorSports {
private:
    string type_of_item;
public:
    Camping(int productID, string name, float price, int amount_in_inventory, string type_of_item)
        : OutdoorSports(productID, name, price, amount_in_inventory), type_of_item(type_of_item) {}

    void display_info() override {
        InventorySystem::display_info();
        cout << "Type: " << type_of_item << endl;
    }
};

// Soccer subclass
class Soccer : public OutdoorSports {
private:
    string type_of_item;
public:
    Soccer(int productID, string name, float price, int amount_in_inventory, string type_of_item)
        : OutdoorSports(productID, name, price, amount_in_inventory), type_of_item(type_of_item) {}

    void display_info() override {
        InventorySystem::display_info();
        cout << "Type: " << type_of_item << endl;
    }
};

// Golf subclass
class Golf : public OutdoorSports {
private:
    string type_of_item;
public:
    Golf(int productID, string name, float price, int amount_in_inventory, string type_of_item)
        : OutdoorSports(productID, name, price, amount_in_inventory), type_of_item(type_of_item) {}

    void display_info() override {
        InventorySystem::display_info();
        cout << "Type: " << type_of_item << endl;
    }
};

// Store class to manage products
class Store {
public:
    vector<InventorySystem*> products;

public:
    // Add product to store
    void add_product(InventorySystem* product) {
        products.push_back(product);
    }

    // Browse all products
    void browse_products() {
        cout << "Available Products:" << endl;
        for (auto& product : products) {
            product->display_info();
        }
    }

    // Search for a product by name
    void search_product(string name) {
        cout << "Searching for '" << name << "':" << endl;
        bool found = false;
        for (auto& product : products) {
            if (product->getProductID() == stoi(name)) {
                product->display_info();
                found = true;
            }
        }
        if (!found) {
            cout << "No products found." << endl;
        }
    }

    // Buy product by productID
    void buy_product(int productID, int quantity) {
        for (auto& product : products) {
            if (product->getProductID() == productID) {
                product->buy_item(quantity);
                return;
            }
        }
        cout << "Product not found." << endl;
    }

    // Restock product by productID
    void restock_product(int productID, int quantity) {
        for (auto& product : products) {
            if (product->getProductID() == productID) {
                product->restock_item(quantity);
                return;
            }
        }
        cout << "Product not found." << endl;
    }
};

// Main function for the User Interface
int main() {
    Store store;

    // Add products to the store
    store.add_product(new Volleyball(101, "Volleyball", 19.99, 10, "Ball"));
    store.add_product(new Volleyball(102, "Volleyball Net", 49.99, 5, "Net"));
    store.add_product(new Basketball(201, "Basketball", 29.99, 8, "Ball"));
    store.add_product(new TableTennis(301, "Table Tennis Paddle", 14.99, 20, "Paddle"));
    store.add_product(new Camping(401, "Camping Tent", 129.99, 3, "Tent"));
    store.add_product(new Soccer(501, "Soccer Cleats", 49.99, 15, "Cleats"));

    // User interface: Basic Menu
    int choice;
    while (true) {
        cout << "\n--- Store Menu ---" << endl;
        cout << "1. Browse Products" << endl;
        cout << "2. Search for Product by ID" << endl;
        cout << "3. Buy Product" << endl;
        cout << "4. Restock Product" << endl;
        cout << "5. Exit" << endl;
        cout << "Enter your choice: ";
        cin >> choice;

        if (choice == 1) {
            store.browse_products();
        } else if (choice == 2) {
            string id;
            cout << "Enter product ID to search: ";
            cin >> id;
            store.search_product(id);
        } else if (choice == 3) {
            int productID, quantity;
            cout << "Enter product ID to buy: ";
            cin >> productID;
            cout << "Enter quantity: ";
            cin >> quantity;
            store.buy_product(productID, quantity);
        } else if (choice == 4) {
            int productID, quantity;
            cout << "Enter product ID to restock: ";
            cin >> productID;
            cout << "Enter quantity to restock: ";
            cin >> quantity;
            store.restock_product(productID, quantity);
        } else if (choice == 5) {
            cout << "Exiting the store..." << endl;
            break;
        } else {
            cout << "Invalid choice. Please try again." << endl;
        }
    }

    // Clean up dynamically allocated memory
    for (auto& product : store.products) {
        delete product;
    }

    return 0;
}

File: HW01.py
class InventorySystem:
    def __init__(self, productID, name, price, amount_in_inventory):
        self.productID = productID
        self.name = name
        self.price = price
        self.amount_in_inventory = amount_in_inventory

    def display_info(self):
        print(f"Product ID: {self.productID}, Name: {self.name}, Price: ${self.price:.2f}, Available: {self.amount_in_inventory}")
    
    def buy_item(self, quantity):
        """Handle the purchase of a certain quantity of an item."""
        if self.amount_in_inventory >= quantity:
            self.amount_in_inventory -= quantity
            print(f"Purchased {quantity} {self.name}(s).")
        else:
            print(f"Not enough stock for {self.name}. Only {self.amount_in_inventory} available.")
    
    def restock_item(self, quantity):
        """Restock the item if necessary."""
        self.amount_in_inventory += quantity
        print(f"Restocked {quantity} {self.name}(s). Now available: {self.amount_in_inventory}.")

class IndoorSports(InventorySystem):
    def __init__(self, productID, name, price, amount_in_inventory):
        super().__init__(productID, name, price, amount_in_inventory)

class OutdoorSports(InventorySystem):
    def __init__(self, productID, name, price, amount_in_inventory):
        super().__init__(productID, name, price, amount_in_inventory)

class Volleyball(IndoorSports):
    def __init__(self, productID, name, price, amount_in_inventory, type_of_item):
        super().__init__(productID, name, price, amount_in_inventory)
        self.type_of_item = type_of_item

    def display_info(self):
        super().display_info()
        print(f"Type: {self.type_of_item}")

class Basketball(IndoorSports):
    def __init__(self, productID, name, price, amount_in_inventory, type_of_item):
        super().__init__(productID, name, price, amount_in_inventory)
        self.type_of_item = type_of_item

    def display_info(self):
        super().display_info()
        print(f"Type: {self.type_of_item}")

class TableTennis(IndoorSports):
    def __init__(self, productID, name, price, amount_in_inventory, type_of_item):
        super().__init__(productID, name, price, amount_in_inventory)
        self.type_of_item = type_of_item

    def display_info(self):
        super().display_info()
        print(f"Type: {self.type_of_item}")

class Camping(OutdoorSports):
    def __init__(self, productID, name, price, amount_in_inventory, type_of_item):
        super().__init__(productID, name, price, amount_in_inventory)
        self.type_of_item = type_of_item

    def display_info(self):
        super().display_info()
        print(f"Type: {self.type_of_item}")

class Soccer(OutdoorSports):
    def __init__(self, productID, name, price, amount_in_inventory, type_of_item):
        super().__init__(productID, name, price, amount_in_inventory)
        self.type_of_item = type_of_item

    def display_info(self):
        super().display_info()
        print(f"Type: {self.type_of_item}")

class Golf(OutdoorSports):
    def __init__(self, productID, name, price, amount_in_inventory, type_of_item):
        super().__init__(productID, name, price, amount_in_inventory)
        self.type_of_item = type_of_item

    def display_info(self):
        super().display_info()
        print(f"Type: {self.type_of_item}")

class Store:
    def __init__(self):
        self.products = []  # List of all products in the store
    
    def add_product(self, product):
        self.products.append(product)
    
    def browse_products(self):
        print("Available Products:")
        for product in self.products:
            product.display_info()
    
    def search_product(self, name):
        print(f"Searching for '{name}':")
        found = False
        for product in self.products:
            if name.lower() in product.name.lower():
                product.display_info()
                found = True
        if not found:
            print("No products found.")
    
    def buy_product(self, productID, quantity):
        for product in self.products:
            if product.productID == productID:
                product.buy_item(quantity)
                return
        print("Product not found.")
    
    def restock_product(self, productID, quantity):
        for product in self.products:
            if product.productID == productID:
                product.restock_item(quantity)
                return
        print("Product not found.")

# Create the store
store = Store()

# Add products to the store
store.add_product(Volleyball(101, "Volleyball", 19.99, 10, "Ball"))
store.add_product(Volleyball(102, "Volleyball Net", 49.99, 5, "Net"))
store.add_product(Basketball(201, "Basketball", 29.99, 8, "Ball"))
store.add_product(TableTennis(301, "Table Tennis Paddle", 14.99, 20, "Paddle"))
store.add_product(Camping(401, "Camping Tent", 129.99, 3, "Tent"))
store.add_product(Soccer(501, "Soccer Cleats", 49.99, 15, "Cleats"))

# User interface: Basic Menu
while True:
    print("\n--- Store Menu ---")
    print("1. Browse Products")
    print("2. Search for Product")
    print("3. Buy Product")
    print("4. Restock Product")
    print("5. Exit")
    choice = input("Enter your choice: ")
    
    if choice == "1":
        store.browse_products()
    elif choice == "2":
        name = input("Enter product name to search: ")
        store.search_product(name)
    elif choice == "3":
        productID = int(input("Enter product ID to buy: "))
        quantity = int(input("Enter quantity: "))
        store.buy_product(productID, quantity)
    elif choice == "4":
        productID = int(input("Enter product ID to restock: "))
        quantity = int(input("Enter quantity to restock: "))
        store.restock_product(productID, quantity)
    elif choice == "5":
        print("Exiting the store...")
        break
    else:
        print("Invalid choice. Please try again.")
